Query compilation breaks using the literal_binds kwarg with an IN operator in the Query on SQLAlchemy 1.4

#### Environment details

  - OS type and version: MacOS Big Sur 11.5.1
  - Python version: 3.7.4
  - pip version: 20.3.3
  - `sqlalchemy-bigquery` version: 0.10.0
  - `sqlalchemy` version: 1.4.22

#### Steps to reproduce

  1.  Create a SQLAlchemy query using an `in_` or `notin_` parameter.
  2. Attempt to compile this query to text using the `compile()` method with `{"literal_binds": True}` in the kwargs. This worked when running with SQLAlchemy version 1.3 and pybigquery version 0.5.0, but does not work with SA 1.4 and pybigquery 0.10.0.

#### Code example

```python
from contextlib import contextmanager

from sqlalchemy import MetaData, Table, create_engine, select

engine = create_engine("bigquery://<your bigquery connection here>")
# Postgres connection, for comparison of proper literal binding and compilation
#engine = create_engine("postgresql://<your postgres connection here>")


@contextmanager
def test_query():
    engine.execute(
        """
        create table test as (
            select 1 as val
            union all select 2
            union all select 3
        )
    """
    )
    val = Table("test", MetaData(bind=engine), autoload=True).c.val
    try:
        yield select(
            [val]
        ).where(val.in_([2]))
    finally:
        engine.execute("drop table test")


with test_query() as q:
    print("Print:")
    print(q)

    print("\nCompile:")
    print(q.compile(engine))

    print("\nLiteral compile:")
    print(q.compile(engine, compile_kwargs={"literal_binds": True}))

    print("\nResult:")
    print(tuple(engine.execute(q)))

```

#### Stack trace
```
Print:
SELECT `test`.`val` 
FROM `test` 
WHERE `test`.`val` IN UNNEST([ [POSTCOMPILE_val_1]:INT64 ])

Compile:
SELECT `test`.`val` 
FROM `test` 
WHERE `test`.`val` IN UNNEST([ [POSTCOMPILE_val_1]:INT64 ])

Literal compile:
Traceback (most recent call last):
  File "/Users/admin/Desktop/test_bq_in.py", line 37, in <module>
    print(q.compile(engine, compile_kwargs={"literal_binds": True}))
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 489, in compile
    return self._compiler(dialect, **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 553, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/pybigquery/sqlalchemy_bigquery.py", line 222, in __init__
    super(BigQueryCompiler, self).__init__(dialect, statement, *args, **kwargs)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 765, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 454, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 489, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 3239, in visit_select
    kwargs,
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 3385, in _compose_select_body
    select._where_criteria, from_linter=from_linter, **kwargs
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 1606, in _generate_delimited_and_list
    return clauses[0]._compiler_dispatch(self, **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 2068, in visit_binary
    return disp(binary, operator_, **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/pybigquery/sqlalchemy_bigquery.py", line 313, in visit_in_op_binary
    self._generate_generic_binary(binary, " IN ", **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/compiler.py", line 2135, in _generate_generic_binary
    self, eager_grouping=eager_grouping, **kw
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/pybigquery/sqlalchemy_bigquery.py", line 437, in visit_bindparam
    assert_(self.__expanded_param(param), f"Unexpected param: {param}")
  File "/Users/admin/anaconda3/envs/sa_migration/lib/python3.7/site-packages/pybigquery/sqlalchemy_bigquery.py", line 63, in assert_
    raise AssertionError(message)
AssertionError: Unexpected param: (2)
```

What we expect to see (From the Postgres running):

```
Print:
SELECT test.val 
FROM test 
WHERE test.val IN ([POSTCOMPILE_val_1])

Compile:
SELECT test.val 
FROM test 
WHERE test.val IN ([POSTCOMPILE_val_1])

Literal compile:
SELECT test.val 
FROM test 
WHERE test.val IN (2)

Result:
((2,),)
```

From the SQLAlchemy documentation (https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#rendering-postcompile-parameters-as-bound-parameters) we expect this to work as setting literal_binds to True should implicitly set render_postcompile to True as well.

--- a/sqlalchemy_bigquery/base.py
+++ b/sqlalchemy_bigquery/base.py
@@ -451,9 +451,10 @@ class BigQueryCompiler(SQLCompiler):
 
         else:
             m = self.__placeholder(param)
-            name, type_ = m.groups()
-            assert_(type_ is None)
-            param = f"%({name}:{bq_type})s"
+            if m:
+                name, type_ = m.groups()
+                assert_(type_ is None)
+                param = f"%({name}:{bq_type})s"
 
         return param
 

--- a/sqlalchemy_bigquery/base.py
+++ b/sqlalchemy_bigquery/base.py
@@ -420,7 +420,7 @@ class BigQueryCompiler(SQLCompiler):
         )
 
         type_ = bindparam.type
-        if isinstance(type_, NullType):
+        if literal_binds or isinstance(type_, NullType):
             return param
 
         if (
@@ -451,10 +451,9 @@ class BigQueryCompiler(SQLCompiler):
 
         else:
             m = self.__placeholder(param)
-            if m:
-                name, type_ = m.groups()
-                assert_(type_ is None)
-                param = f"%({name}:{bq_type})s"
+            name, type_ = m.groups()
+            assert_(type_ is None)
+            param = f"%({name}:{bq_type})s"
 
         return param
 

--- a/testing/constraints-3.6.txt
+++ b/testing/constraints-3.6.txt
@@ -6,5 +6,5 @@
 # e.g., if setup.py has "foo >= 1.14.0, < 2.0.0dev",
 sqlalchemy==1.2.0
 google-auth==1.25.0
-google-cloud-bigquery==2.19.0
+google-cloud-bigquery==2.24.1
 google-api-core==1.30.0

--- a/tests/system/test_sqlalchemy_bigquery.py
+++ b/tests/system/test_sqlalchemy_bigquery.py
@@ -301,7 +301,7 @@ def test_record_content_from_raw_queries(engine, bigquery_dataset):
 
 
 def test_content_from_reflect(engine, table_one_row):
-    rows = table_one_row.select().execute().fetchall()
+    rows = table_one_row.select(use_labels=True).execute().fetchall()
     assert list(rows[0]) == ONE_ROW_CONTENTS_EXPANDED
 
 
@@ -505,21 +505,21 @@ def test_querying_wildcard_tables(engine):
 def test_dml(engine, session, table_dml):
     # test insert
     engine.execute(table_dml.insert(ONE_ROW_CONTENTS_DML))
-    result = table_dml.select().execute().fetchall()
+    result = table_dml.select(use_labels=True).execute().fetchall()
     assert len(result) == 1
 
     # test update
     session.query(table_dml).filter(table_dml.c.string == "test").update(
         {"string": "updated_row"}, synchronize_session=False
     )
-    updated_result = table_dml.select().execute().fetchone()
+    updated_result = table_dml.select(use_labels=True).execute().fetchone()
     assert updated_result[table_dml.c.string] == "updated_row"
 
     # test delete
     session.query(table_dml).filter(table_dml.c.string == "updated_row").delete(
         synchronize_session=False
     )
-    result = table_dml.select().execute().fetchall()
+    result = table_dml.select(use_labels=True).execute().fetchall()
     assert len(result) == 0
 
 

--- a/tests/unit/test_compliance.py
+++ b/tests/unit/test_compliance.py
@@ -215,4 +215,4 @@ def test_cast_type_decorator(faux_conn, last_query):
 
     t = setup_table(faux_conn, "t", Column("x", StringAsInt()))
     faux_conn.execute(t.insert(), [{"x": x} for x in [1, 2, 3]])
-    last_query('INSERT INTO `t` (`x`) VALUES (CAST(%(x:STRING)s AS STRING))', {'x': 3})
+    last_query("INSERT INTO `t` (`x`) VALUES (CAST(%(x:STRING)s AS STRING))", {"x": 3})

--- a/tests/unit/test_compliance.py
+++ b/tests/unit/test_compliance.py
@@ -200,3 +200,19 @@ def test_group_by_composed(faux_conn):
         select([sqlalchemy.func.count(table.c.id), expr]).group_by(expr).order_by(expr)
     )
     assert_result(faux_conn, stmt, [(1, 3), (1, 5), (1, 7)])
+
+
+def test_cast_type_decorator(faux_conn, last_query):
+    # [artial dup of:
+    #   sqlalchemy.testing.suite.test_types.CastTypeDecoratorTest.test_special_type
+    # That test failes without code that's otherwise not covered by the unit tests.
+
+    class StringAsInt(sqlalchemy.TypeDecorator):
+        impl = sqlalchemy.String(50)
+
+        def bind_expression(self, col):
+            return sqlalchemy.cast(col, String(50))
+
+    t = setup_table(faux_conn, "t", Column("x", StringAsInt()))
+    faux_conn.execute(t.insert(), [{"x": x} for x in [1, 2, 3]])
+    last_query('INSERT INTO `t` (`x`) VALUES (CAST(%(x:STRING)s AS STRING))', {'x': 3})

--- a/tests/unit/test_select.py
+++ b/tests/unit/test_select.py
@@ -356,3 +356,17 @@ def test_select_notin_param_empty(faux_conn):
         else "SELECT (%(param_1:INT64)s NOT IN UNNEST([  ])) AS `anon_1`",
         {"param_1": 1},
     )
+
+
+def test_literal_binds_kwarg_with_an_IN_operator_252(faux_conn):
+    table = setup_table(faux_conn, "test", sqlalchemy.Column("val", sqlalchemy.Integer),
+                        initial_data=[dict(val=i) for i in range(3)])
+    q = sqlalchemy.select([table.c.val]).where(table.c.val.in_([2]))
+
+    def nstr(q):
+        return ' '.join(str(q).strip().split())
+
+    assert nstr(q.compile(faux_conn.engine,
+                          compile_kwargs={"literal_binds": True}
+                          )
+                ) == 'SELECT `test`.`val` FROM `test` WHERE `test`.`val` IN (2)'

--- a/tests/unit/test_select.py
+++ b/tests/unit/test_select.py
@@ -359,14 +359,18 @@ def test_select_notin_param_empty(faux_conn):
 
 
 def test_literal_binds_kwarg_with_an_IN_operator_252(faux_conn):
-    table = setup_table(faux_conn, "test", sqlalchemy.Column("val", sqlalchemy.Integer),
-                        initial_data=[dict(val=i) for i in range(3)])
+    table = setup_table(
+        faux_conn,
+        "test",
+        sqlalchemy.Column("val", sqlalchemy.Integer),
+        initial_data=[dict(val=i) for i in range(3)],
+    )
     q = sqlalchemy.select([table.c.val]).where(table.c.val.in_([2]))
 
     def nstr(q):
-        return ' '.join(str(q).strip().split())
+        return " ".join(str(q).strip().split())
 
-    assert nstr(q.compile(faux_conn.engine,
-                          compile_kwargs={"literal_binds": True}
-                          )
-                ) == 'SELECT `test`.`val` FROM `test` WHERE `test`.`val` IN (2)'
+    assert (
+        nstr(q.compile(faux_conn.engine, compile_kwargs={"literal_binds": True}))
+        == "SELECT `test`.`val` FROM `test` WHERE `test`.`val` IN (2)"
+    )

